public/app/features/dashboard-scene/scene/layouts-shared/paste.ts

```diff
@@ -0,0 +1,137 @@
+} from '@grafana/schema/dist/esm/schema/dashboard/v2alpha0';
+import { LS_PANEL_COPY_KEY, LS_ROW_COPY_KEY, LS_TAB_COPY_KEY } from 'app/core/constants';
+import store from 'app/core/store';
+
+import { NewObjectAddedToCanvasEvent } from '../../edit-pane/shared';
+import { deserializeRow } from '../../serialization/layoutSerializers/RowsLayoutSerializer';
+import { deserializeTab } from '../../serialization/layoutSerializers/TabsLayoutSerializer';
+import { dashboardSceneGraph } from '../../utils/dashboardSceneGraph';
+import { DashboardScene } from '../DashboardScene';
+import { RowItem } from '../layout-rows/RowItem';
+import { RowsLayoutManager } from '../layout-rows/RowsLayoutManager';
+import { TabItem } from '../layout-tabs/TabItem';
+import { TabsLayoutManager } from '../layout-tabs/TabsLayoutManager';
+import { DashboardLayoutManager } from '../types/DashboardLayoutManager';
+import { isLayoutParent } from '../types/LayoutParent';
+
+export function clearClipboard() {
+  store.delete(LS_PANEL_COPY_KEY);
+  store.delete(LS_ROW_COPY_KEY);
+  store.delete(LS_TAB_COPY_KEY);
+}
+
+export interface RowStore {
+  elements: DashboardV2Spec['elements'];
+  row: RowsLayoutRowKind;
+}
+
+export interface TabStore {
+  elements: DashboardV2Spec['elements'];
+  tab: TabsLayoutTabKind;
+}
+
+export function pasteTabTo(layout: DashboardLayoutManager, tab: TabItem): TabItem {
+  const layoutParent = layout.parent!;
+  if (!isLayoutParent(layoutParent)) {
+    throw new Error('Parent layout is not a LayoutParent');
+  }
+
+  if (layout instanceof TabsLayoutManager) {
+    return layout.addNewTab(tab);
+  }
+
+  // Create new tabs layout and wrap the current layout in the first tab
+
+  if (layoutParent.getLayout().isEmpty()) {
+    const tabsLayout = TabsLayoutManager.createEmpty();
+    tabsLayout.state.tabs[0].setState(tab.state);
+    layoutParent.switchLayout(tabsLayout);
+    return tabsLayout.state.tabs[0];
+  }
+
+  const tabsLayout = TabsLayoutManager.createEmpty();
+  tabsLayout.state.tabs[0].setState({ layout: layout.clone() });
+
+  layoutParent.switchLayout(tabsLayout);
+
+  const newTab = tabsLayout.state.tabs[0];
+  layout.publishEvent(new NewObjectAddedToCanvasEvent(newTab), true);
+
+  return tabsLayout.addNewTab(tab);
+}
+
+export function pasteRowTo(layout: DashboardLayoutManager, row: RowItem): RowItem {
+  if (layout instanceof RowsLayoutManager) {
+    return layout.addNewRow(row);
+  }
+
+  if (layout instanceof TabsLayoutManager) {
+    const currentTab = layout.getCurrentTab();
+    return pasteRowTo(currentTab.state.layout, row);
+  }
+
+  const layoutParent = layout.parent!;
+  if (!isLayoutParent(layoutParent)) {
+    throw new Error('Parent layout is not a LayoutParent');
+  }
+
+  if (layoutParent.getLayout().isEmpty()) {
+    const rowsLayout = RowsLayoutManager.createEmpty();
+    rowsLayout.state.rows[0].setState(row.state);
+    layoutParent.switchLayout(rowsLayout);
+    return rowsLayout.state.rows[0];
```

Comment:
I can't replicate this scenario, I can only paste a row to a rows layout, so no need to switch any layout. 

Same with the tab scenario. Which I think is fine, so if I understand things correctly we can remove this condition (and all the isEmpty functions)

Reply:
Yeah. I don't really like the addition of is empty to everything. 

You can paste a row to a tab layout using the add pane.

The reason it's there is because when we paste a row to a tab that does not have a rows layout, we will switch it to a rows layout. When switching to the rows layout we encapsulate the grid in a newly created row, then add the pasted row as the second row.
This looks a bit weird when we paste to a newly created and empty tab. It will encapsulate the empty grid in a row and then add the pasted row.

Pasting from the add pane creates a lot of complications. But it also feels a bit awkward to first create a rows layout to paste a row.

https://github.com/user-attachments/assets/d900391a-0d02-4e63-a6be-e01ba7543bf0

Reply:
removing paste from add pane removed this complication